File: engine-core/src/schema.rs
//! The on-chain schema registry.
//!
//! Component layouts are registered on-chain, which gives the engine reflection
//! (read any Component's bytes given its schema) and enables cross-world
//! composability: world B can reference a Component defined in world A because both
//! speak the same schema.
//!
//! v1 schemas are fixed-size scalar records. A constant size per Component keeps
//! account rent, the compute-budget math in [`crate::tick`], and the encode/decode
//! in [`crate::component`] deterministic. Dynamic/variable fields are a follow-up.

use crate::error::EngineError;
use crate::ids::ComponentId;
use std::collections::BTreeMap;

/// The scalar field types a Component may contain. Each has a fixed byte width and
/// a fixed little-endian encoding shared with the TypeScript SDK.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum FieldType {
    U8,
    U16,
    U32,
    U64,
    I64,
    /// A boolean stored as a single `0`/`1` byte.
    Bool,
    /// A 32-byte Solana public key.
    Pubkey,
    /// An [`crate::ids::EntityId`] reference (u64). Distinguished from `U64` so the
    /// indexer and cross-world references know it points at another entity.
    Entity,
    /// A fixed-length byte blob of `n` bytes - for packed data a System consumes
    /// wholesale (e.g. a settled pathfinding route, an inventory bitmap).
    Bytes(u16),
}

impl FieldType {
    /// Fixed encoded width in bytes.
    pub const fn size(self) -> usize {
        match self {
            FieldType::U8 | FieldType::Bool => 1,
            FieldType::U16 => 2,
            FieldType::U32 => 4,
            FieldType::U64 | FieldType::I64 | FieldType::Entity => 8,
            FieldType::Pubkey => 32,
            FieldType::Bytes(n) => n as usize,
        }
    }

    /// Stable wire discriminant, shared with the TypeScript SDK. Not the same as
    /// an `as u8` cast: it must never shift as variants are added.
    pub const fn tag(self) -> u8 {
        match self {
            FieldType::U8 => 1,
            FieldType::U16 => 2,
            FieldType::U32 => 3,
            FieldType::U64 => 4,
            FieldType::I64 => 5,
            FieldType::Bool => 6,
            FieldType::Pubkey => 7,
            FieldType::Entity => 8,
            FieldType::Bytes(_) => 9,
        }
    }
}

/// One named field within a Component schema.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Field {
    pub name: String,
    pub ty: FieldType,
}

impl Field {
    pub fn new(name: impl Into<String>, ty: FieldType) -> Self {
        Self {
            name: name.into(),
            ty,
        }
    }
}

/// A registered Component layout.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct ComponentSchema {
    pub id: ComponentId,
    pub name: String,
    pub fields: Vec<Field>,
}

impl ComponentSchema {
    pub fn new(id: ComponentId, name: impl Into<String>, fields: Vec<Field>) -> Self {
        Self {
            id,
            name: name.into(),
            fields,
        }
    }

    /// Total fixed size of one record of this schema.
    pub fn size(&self) -> usize {
        self.fields.iter().map(|f| f.ty.size()).sum()
    }

    /// Byte offset and type of a field by name, if present.
    pub fn field(&self, name: &str) -> Option<(usize, FieldType)> {
        let mut offset = 0usize;
        for f in &self.fields {
            if f.name == name {
                return Some((offset, f.ty));
            }
            offset += f.ty.size();
        }
        None
    }

    /// A stable digest of the layout. Two schemas with the same digest are
    /// byte-compatible, and thus cross-world referenceable.
    pub fn layout_hash(&self) -> u64 {
        let mut h = crate::hash::Hasher::new();
        h.write(b"loom:schema").write_u32(self.id);
        h.write(self.name.as_bytes());
        for f in &self.fields {
            h.write_u8(0xff); // field separator
            h.write(f.name.as_bytes());
            h.write_u8(f.ty.tag());
            if let FieldType::Bytes(n) = f.ty {
                h.write(&n.to_le_bytes());
            }
        }
        h.finish()
    }
}

/// A world's registry of Component schemas, keyed by id and ordered for
/// deterministic iteration and hashing.
#[derive(Clone, Default, Debug)]
pub struct SchemaRegistry {
    schemas: BTreeMap<ComponentId, ComponentSchema>,
    next_id: ComponentId,
}

impl SchemaRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Register a schema under the next free id and return that id.
    pub fn register(&mut self, name: impl Into<String>, fields: Vec<Field>) -> ComponentId {
        let id = self.next_id;
        self.next_id += 1;
        self.schemas
            .insert(id, ComponentSchema::new(id, name, fields));
        id
    }

    /// Register a schema with an explicit id (used when replaying on-chain state).
    pub fn register_with_id(
        &mut self,
        id: ComponentId,
        name: impl Into<String>,
        fields: Vec<Field>,
    ) -> Result<(), EngineError> {
        if self.schemas.contains_key(&id) {
            return Err(EngineError::DuplicateComponent(id));
        }
        self.schemas
            .insert(id, ComponentSchema::new(id, name, fields));
        self.next_id = self.next_id.max(id + 1);
        Ok(())
    }

    pub fn get(&self, id: ComponentId) -> Result<&ComponentSchema, EngineError> {
        self.schemas
            .get(&id)
            .ok_or(EngineError::UnknownComponent(id))
    }

    pub fn contains(&self, id: ComponentId) -> bool {
        self.schemas.contains_key(&id)
    }

    pub fn len(&self) -> usize {
        self.schemas.len()
    }

    pub fn is_empty(&self) -> bool {
        self.schemas.is_empty()
    }

    /// Iterate schemas in ascending id order.
    pub fn iter(&self) -> impl Iterator<Item = &ComponentSchema> {
        self.schemas.values()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn position() -> Vec<Field> {
        vec![
            Field::new("x", FieldType::I64),
            Field::new("y", FieldType::I64),
        ]
    }

    #[test]
    fn offsets_and_size() {
        let s = ComponentSchema::new(0, "Position", position());
        assert_eq!(s.size(), 16);
        assert_eq!(s.field("x"), Some((0, FieldType::I64)));
        assert_eq!(s.field("y"), Some((8, FieldType::I64)));
        assert_eq!(s.field("z"), None);
    }

    #[test]
    fn registry_allocates_ids() {
        let mut r = SchemaRegistry::new();
        let pos = r.register("Position", position());
        let vel = r.register("Velocity", position());
        assert_eq!(pos, 0);
        assert_eq!(vel, 1);
        assert_eq!(r.len(), 2);
        assert_eq!(r.get(pos).unwrap().name, "Position");
    }

    #[test]
    fn layout_hash_is_structural() {
        let a = ComponentSchema::new(0, "Position", position());
        let b = ComponentSchema::new(0, "Position", position());
        let c = ComponentSchema::new(
            0,
            "Position",
            vec![Field::new("x", FieldType::I64), Field::new("y", FieldType::U64)],
        );
        assert_eq!(a.layout_hash(), b.layout_hash());
        assert_ne!(a.layout_hash(), c.layout_hash());
    }
}

File: sdk/src/schema.ts
// Schema registry. Components are fixed-size scalar records.

import { Hasher, utf8 } from "./hash.ts";
import { EngineError } from "./errors.ts";

/** Scalar field types. `{ bytes: n }` is a fixed-length blob of `n` bytes. */
export type FieldType =
  | "u8"
  | "u16"
  | "u32"
  | "u64"
  | "i64"
  | "bool"
  | "pubkey"
  | "entity"
  | { bytes: number };

/** Fixed encoded width of a field type, in bytes. */
export function fieldSize(ty: FieldType): number {
  if (typeof ty === "object") return ty.bytes;
  switch (ty) {
    case "u8":
    case "bool":
      return 1;
    case "u16":
      return 2;
    case "u32":
      return 4;
    case "u64":
    case "i64":
    case "entity":
      return 8;
    case "pubkey":
      return 32;
  }
}

/** Stable wire discriminant for a field type. */
export function fieldTag(ty: FieldType): number {
  if (typeof ty === "object") return 9; // bytes
  switch (ty) {
    case "u8":
      return 1;
    case "u16":
      return 2;
    case "u32":
      return 3;
    case "u64":
      return 4;
    case "i64":
      return 5;
    case "bool":
      return 6;
    case "pubkey":
      return 7;
    case "entity":
      return 8;
  }
}

export interface Field {
  name: string;
  ty: FieldType;
}

export function field(name: string, ty: FieldType): Field {
  return { name, ty };
}

export class ComponentSchema {
  readonly id: number;
  readonly name: string;
  readonly fields: Field[];

  constructor(id: number, name: string, fields: Field[]) {
    this.id = id;
    this.name = name;
    this.fields = fields;
  }

  /** Total fixed size of one record. */
  size(): number {
    return this.fields.reduce((acc, f) => acc + fieldSize(f.ty), 0);
  }

  /** Byte offset and type of a field by name. */
  field(name: string): { offset: number; ty: FieldType } | undefined {
    let offset = 0;
    for (const f of this.fields) {
      if (f.name === name) return { offset, ty: f.ty };
      offset += fieldSize(f.ty);
    }
    return undefined;
  }

  /** Structural layout digest. */
  layoutHash(): bigint {
    const h = new Hasher();
    h.write(utf8("loom:schema")).writeU32(this.id).write(utf8(this.name));
    for (const f of this.fields) {
      h.writeU8(0xff).write(utf8(f.name)).writeU8(fieldTag(f.ty));
      if (typeof f.ty === "object") h.writeU16(f.ty.bytes);
    }
    return h.finish();
  }
}

/** A world's registry of Component schemas, ordered by id. */
export class SchemaRegistry {
  private schemas = new Map<number, ComponentSchema>();
  private nextId = 0;

  register(name: string, fields: Field[]): number {
    const id = this.nextId++;
    this.schemas.set(id, new ComponentSchema(id, name, fields));
    return id;
  }

  registerWithId(id: number, name: string, fields: Field[]): void {
    if (this.schemas.has(id)) {
      throw new EngineError("DuplicateComponent", `component ${id} already registered`, { id });
    }
    this.schemas.set(id, new ComponentSchema(id, name, fields));
    this.nextId = Math.max(this.nextId, id + 1);
  }

  get(id: number): ComponentSchema {
    const s = this.schemas.get(id);
    if (!s) throw new EngineError("UnknownComponent", `unknown component ${id}`, { id });
    return s;
  }

  has(id: number): boolean {
    return this.schemas.has(id);
  }

  get length(): number {
    return this.schemas.size;
  }

  /** Schemas in ascending id order. */
  iter(): ComponentSchema[] {
    return [...this.schemas.keys()].sort((a, b) => a - b).map((id) => this.schemas.get(id)!);
  }
}
